src/app/panel/auth/SignUp.tsx

```diff
@@ -0,0 +1,90 @@
+import React, { useState } from 'react';
+import { TextInput, Button } from '@tremor/react';
+import { APIService } from '@/services/api';
+
+const SignUp: React.FC = () => {
+  const [username, setUsername] = useState<string>('');
+  const [email, setEmail] = useState<string>('');
+  const [password, setPassword] = useState<string>('');
+  const [confirmPassword, setConfirmPassword] = useState<string>('');
+  const [errorMessage, setErrorMessage] = useState<string | null>('');
+  const [isFormValid, setIsFormValid] = useState<boolean>(false);
+
+  const handleSubmit = (e: React.FormEvent) => {
+    e.preventDefault();
+    APIService.signUp(username, email, password);
+  };
+
+  const onUsernameChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    e.target.value = e.target.value.replace(/[^a-zA-Z0-9]/g, '');
+    setUsername(e.target.value);
+    validateForm();
+  }
+
+  const onEmailChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    e.target.value = e.target.value.replace(/[^a-zA-Z0-9]/g, '');
+    setEmail(e.target.value);
+    validateForm();
+  } 
+
+  const onPasswordChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    e.target.value = e.target.value.replace(/[^a-zA-Z0-9]/g, '');
+    setPassword(e.target.value);
+    validateForm();
+  }
+
+  const onConfirmPasswordChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    e.target.value = e.target.value.replace(/[^a-zA-Z0-9]/g, '');
+    setConfirmPassword(e.target.value);
+    validateForm();
+  }
+
+  const validateForm = (): void => {
```

Comment:
Why are we redefining this function? I think there can be a common service that does this which both SignIn and SignUp could use.

Reply:
There is only 2 forms and both have different inputs, I think that would be an over-generalisation